### ISSUE TITLE
Moving average ignored a value of inf/nan

### ISSUE THREAD

**Issue by emjay73** [dbolya/yolact#186]
Well, I thought this problem was fixed after several attempts, but it was not. :(

![Screenshot from 2019-10-22 15-30-47](https://user-images.githubusercontent.com/36818801/67262829-1cd43900-f4e1-11e9-965a-80eb11ccbbcf.png)
![Screenshot from 2019-10-22 14-27-15](https://user-images.githubusercontent.com/36818801/67262839-29589180-f4e1-11e9-873a-011c93cdec90.png)
![Screenshot from 2019-10-22 15-56-44](https://user-images.githubusercontent.com/36818801/67263943-93bf0100-f4e4-11e9-87e7-60a78fa4d326.png)
![Screenshot from 2019-10-22 16-16-36](https://user-images.githubusercontent.com/36818801/67264899-5ad45b80-f4e7-11e9-88f9-05936df5cdcb.png)

I tried it several times and it dies eventually with a loss explosion.
I used default settings with following arguments,
python train.py --config=yolact_base_config

**Comment by dbolya** [comment]
Wait, are you running this across all 4 GPUs? Try first running
```
export CUDA_VISIBLE_DEVICES=0
```
first. Since YOLACT uses batch norm, having a per-gpu batch size of under 6 would be big no-no (leading to an unstable mess). The default batch size is 8 and stretching that across 4 GPUs would make the per-gpu batch size 2, which is a bad time for everyone involved.

**Comment by emjay73** [comment]
That makes sense.
With a single GPU0, it seems it's working just fine.

And I have an additional question.
As far as I know, batch size itself could affect evaluation accuracy, isn't it?
Does ((batch size = 8*4) & ( 4 GPUs )) guarantee the same effect as ((batch size = 8) & ( single GPU ))?
The same question goes with iteration * 4 and lr*4. (especially lr.)

**Comment by dbolya** [comment]
I'm actually testing just that right now (on pascal with resnet50 so might change a little, but should be the same). So it looks like if you follow the steps I posted in the last thread:
```
batch_size *= num_gpus
lr *= num_gpus
iterations /= num_gpus
```
This is following the advice from this paper: https://arxiv.org/abs/1706.02677

Otherwise, you're right, the evaluation performance isn't the same. Pretty soon I'll have these adjustments performed automatically among other multi-gpu optimizations. I swear, I'll push these before ICCV if it's the last thing I do.

My current stats for 2080 Tis on pascal with resnet50 are 14 hours for 1 gpu and 5 hours for 4 gpus. Before I finish, I want to solve the issue where randomly you'll get an inf loss for a batch (when the loss hasn't exploded). This was fine for batch size=8 because I would just discard the batch which wouldn't matter much, but now with batch_size=44 it's a little bit worse to discard the whole batch...

Oh and yeah I want to get 4 GPUs to run at 4x the speed of 1 GPU, but we'll see if I get there.

**Comment by emjay73** [comment]
I hope you can get there :)
-with fervent support

**Comment by dbolya** [comment]
I'll let you know if I do, but for now I'll close the issue :^ )
